types/sharedb/lib/client.d.ts

```diff
@@ -29,6 +29,30 @@ export class Connection {
     fetchSnapshotByTimestamp(collection: string, id: string, timestamp: number, callback: (error: Error, snapshot: ShareDB.Snapshot) => void): void;
     getPresence(channel: string): Presence;
     getDocPresence(collection: string, id: string): Presence;
+
+    /**
+     * Returns whether anything in this client is either:
+     * - In-flight, waiting on a response from the server
+     * - Pending (locally queued)
+     */
+    hasPending(): boolean;
+
+    /**
+     * Invokes the callback once nothing on this client is in-flight or pending.
+     *
+     * @see hasPending
+     */
+    whenNothingPending(callback: () => void): void;
```

Comment:
This callback [can accept an error](https://github.com/share/sharedb/blob/07f8bd98e5cdae37e925d4c0c709a1c3cbbca5e0/lib/client/doc.js#L126). You can use the existing `Callback` type this module already has:

```suggestion
    whenNothingPending(callback: ShareDB.Callback): void;
```

Reply:
That's a link to `Doc#whenNothingPending`.

`Connection#whenNothingPending` is always called with no arguments:
https://github.com/share/sharedb/blob/07f8bd98e5cdae37e925d4c0c709a1c3cbbca5e0/lib/client/connection.js#L607-L608

That's because there could be many pending operations that need to be resolved before `Connection#whenNothingPending` will call back, and potentially many errors. Makes more sense to handle each error on the specific object that the error is for.

Reply:
Aha my bad